sys/include/net/dhcpv6.h

```diff
@@ -0,0 +1,305 @@
+#define DHCPV6_REPLY                (7U)    /**< REPLY */
+/** @ } */
+
+/**
+ * @name    DHCPv6 option codes
+ * @anchor  net_dhcp6_opt_codes
+ * @see [draft-ietf-dhc-rfc3315bis-12, section 21]
+ *      (https://tools.ietf.org/html/draft-ietf-dhc-rfc3315bis-12#section-21)
+ * @{
+ */
+#define DHCPV6_OPT_CID              (1U)    /**< client identifier option */
+#define DHCPV6_OPT_SID              (2U)    /**< server identifier option */
+#define DHCPV6_OPT_ORO              (6U)    /**< option request option */
+#define DHCPV6_OPT_PREF             (7U)    /**< preference option */
+#define DHCPV6_OPT_ELAPSED_TIME     (8U)    /**< elapsed time option */
+#define DHCPV6_OPT_STATUS           (13U)   /**< status code option */
+#define DHCPV6_OPT_IA_PD            (25U)   /**< identity association for prefix
+                                             *   delegation (IA_PD) option */
+#define DHCPV6_OPT_IAPFX            (26U)   /**< IA prefix option */
+#define DHCPV6_OPT_SMR              (82U)   /**< SOL_MAX_RT option */
+/** @} */
+
+/**
+ * @name    DHCPv6 status codes
+ * @anchor  net_dhcp6_status_codes
+ * @see [draft-ietf-dhc-rfc3315bis-12, section 21.13]
+ *      (https://tools.ietf.org/html/draft-ietf-dhc-rfc3315bis-12#section-21.13)
+ * @{
+ */
+#define DHCPV6_STATUS_SUCCESS       (0U)    /**< Success */
+/** @} */
+
+/**
+ * @name    DHCPv6 transmission and retransmission parameters
+ * @see [draft-ietf-dhc-rfc3315bis-12, section 7.6]
+ *      (https://tools.ietf.org/html/draft-ietf-dhc-rfc3315bis-12#section-7.6)
+ * @{
+ */
+#define DHCPV6_SOL_MAX_DELAY        (1U)    /**< SOL_MAX_DELAY (in sec) */
+#define DHCPV6_SOL_TIMEOUT          (1U)    /**< SOL_TIMEOUT (in sec) */
+#define DHCPV6_SOL_MAX_RT           (3600U) /**< SOL_MAX_RT (in sec) */
+
+#define DHCPV6_REQ_TIMEOUT          (1U)    /**< REQ_TIMEOUT (in sec) */
+#define DHCPV6_REQ_MAX_RT           (30U)   /**< REQ_MAX_RT (in sec) */
+#define DHCPV6_REQ_MAX_RC           (10U)   /**< REQ_MAX_RC */
+
+#define DHCPV6_REN_TIMEOUT          (10U)   /**< REN_TIMEOUT (in sec) */
+#define DHCPV6_REN_MAX_RT           (600U)  /**< REN_MAX_RT (in sec) */
+
+#define DHCPV6_REB_TIMEOUT          (10U)   /**< REB_TIMEOUT (in sec) */
+#define DHCPV6_REB_MAX_RT           (600U)  /**< REB_MAX_RT (in sec) */
+/** @} */
+
+#define DHCPV6_DUID_MAX_LEN         (128U)  /**< maximum length of DUID */
+
+/**
+ * @name DHCPv6 unique identifier (DUID) types
+ * @see [draft-ietf-dhc-rfc3315bis-12, section 11.1]
+ *      (https://tools.ietf.org/html/draft-ietf-dhc-rfc3315bis-12#section-11.1)
+ * @{
+ */
+#define DHCPV6_DUID_TYPE_L2         (3U)    /**< Link-layer address */
+/** @} */
+
+/**
+ * @name DHCPv6 message formats
+ * @{
+ */
+/**
+ * @brief   Client/Server message header
+ * @see [draft-ietf-dhc-rfc3315bis-12, section 8]
+ *      (https://tools.ietf.org/html/draft-ietf-dhc-rfc3315bis-12#section-8)
+ */
+typedef struct __attribute__((packed)) {
```

Comment:
I suggest keeping the implementation details specific to the implementation.

Reply:
This is not an implementation, but a protocol detail...

Reply:
Say we implement a relay agent or server in the future, they will need this info, too, while being independent from the client.

Reply:
Well, I usually implement protocols without using network_uintX_t types. Thus their usage is implementation dependent. Also, defining packed header types which can be casted implies exactly that is done, which is, while standard procedure, implementation dependent. -> please move out of the global sys/net/dhcpv6.h

Reply:
That's news to me... I never heard about this convention and precedence ([here](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/ipv6/hdr.h, [here](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/udp.h), [here](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/ethernet/hdr.h) and other places) seem to provide backing for that.

Reply:
Well, never too late to do the right thing.

Reply:
Let's agree to disagree that I have my style of defining protocol specifics (having easy to interpret code from an RFC reader's stand-point and having build-in compiler warnings when forgetting about byte-order) and you have yours (reducing dependencies to native RIOT headers?) and both are neither wrong nor right.

Reply:
Please let's not do a you-vs-me thing here, we're both trying to improve RIOT.

I'm not arguing against a certain style of implementing a protocol, just where to put the stuff. In this case, I'm asking to keep the implementation style with the implementation.

Reply:
But this isn't implementation-specific IMHO. This is the definition of the message format which is an integral part of the protocol definition (aka this header).

Reply:
Maybe. And it looks nice (if you agree with the style). But the questions should be: 

- Where is it used?
- Is it needed in the global header space?

Seems to me you're trying to convert RFC into C header and put it in a prominent place. Why?

IMO, most of the types and defines should stay within some dhcp*.c.
Or do you think the dhcpv6_hdr_t will be re-used by a different implementation?

Reply:
> Or do you think the dhcpv6_hdr_t will be re-used by a different implementation?

Actually, yes! E.g. will need relay agents at some point if we want to have multi-hop address assignments in a WPAN. If the message formats were not defined somewhere central this would lead to code duplication (if only by definitions). Also for the case concurring implementations to `dhcpv6_client` (maybe someone doesn't like my code ;-)): why shouldn't they? These definitions already take one of the more excruciating, repetitive, and boring tasks when implementing (parsing the information out of an RFC and molding them in C) a protocol and just provides all (or in this case here, most of) the message definitions for free.

Reply:
> E.g. will need relay agents at some point if we want to have multi-hop address assignments in a WPAN.

At that point we could reconsider moving the defines "up".

> Also for the case concurring implementations to dhcpv6_client (maybe someone doesn't like my code ;-)): why shouldn't they?

Because if they do things the same, they don't have much right to share the same repo. If they do things differently, that might include how headers & definitions look like.

I have some experience in implementing things differently than done in gnrc, and trust me, it sucks to have the headers in the way.

Reply:
> I have some experience in implementing things differently than done in gnrc, and trust me, it sucks to have the headers in the way.

As it does to have non-implementation specific stuff in the implementation specific header (two examples are `GCOAP_PORT` and `EMCUTE_PORT` ;-)). However, `network_uintX_t` has nothing to do with GNRC, but with RIOT so why do you have problems with something RIOT-specific when working with RIOT?